### PR TITLE
Port nballerina semtype tests

### LINKED ISSUE #34629: Port SemType Tests from nBallerina to lang repo Tests
$Subject.
No of people: 2

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #34629

### CODE REVIEW

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                continue;
+            }
+            files.add(file.getAbsolutePath());
+        }
+        return files.toArray();
+    }
+
+    @Test(dataProvider = "filePathProvider")
+    public void testSubTypeRelationship(String filePath) throws IOException {
+        List<String> actualRelations = actualSubTypeRelations(filePath);
```
Comment (rdhananjaya) [inline tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:73]: I have a feeling that `actualRelations` and `expectedRelations` should be `Set<String>`.

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                String n2 = td2.name.getValue();
+
+                checkSubTypeRelation(t1, t2, n1, n2, types, relations);
+                checkSubTypeRelation(t2, t1, n2, n1, types, relations);
+            }
+        }
+        return relations;
+    }
+
+    private void checkSubTypeRelation(BType sourceType, BType targetType, String sourceTypeName,
```
Comment (rdhananjaya) [inline tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:112]:
```suggestion
    private void computeSubTypeRelation(BType sourceType, BType targetType, String sourceTypeName,
```

Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java]: Changed

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java (hunk 1 above)
Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:73]: @rdhananjaya, is there any specific reason to do this?

Comment (rdhananjaya) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:73]: We only need set difference between expected and calculated relations.

Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java]: Changed

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            relations.add(relation(sourceTypeName, targetTypeName));
+        }
+    }
+
+    private Set<String> expectedSubTypeRelations(String filePath) throws IOException {
+        Set<String> relations = new HashSet<>();
+        List<String> lines = Files.readAllLines(resolvePath(filePath));
+        for (String line : lines) {
+            // This check can be changed based on other notations.
+            if (line.startsWith(COMMENT) && line.contains(SUBTYPE_SYMBOL)) {
```
Comment (pcnfernando) [inline tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:124]: Can we improve the symbol validation? eg: having two spaces between the comment and the symbol will brake this

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                bTypeDefinitionSymbols.add((BTypeDefinitionSymbol) bSymbol);
+            }
+        }
+
+        Types types = Types.getInstance(new CompilerContext());
+        Set<String> relations = new HashSet<>();
+        int len = bTypeDefinitionSymbols.size();
+        for (int i = 0; i < len; i++) {
+            for (int j = i + 1; j < len; j++) {
+                BTypeDefinitionSymbol td1 = bTypeDefinitionSymbols.get(i);
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java]: Changed

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        BLangPackage bLangPackage = (BLangPackage) compileResult.getAST();
+        List<BTypeDefinitionSymbol> bTypeDefinitionSymbols = new ArrayList<>();
+        for (Scope.ScopeEntry value : bLangPackage.symbol.scope.entries.values()) {
+            BSymbol bSymbol = value.symbol;
+            if (bSymbol.kind == SymbolKind.TYPE_DEF && bSymbol.origin == SymbolOrigin.SOURCE) {
+                bTypeDefinitionSymbols.add((BTypeDefinitionSymbol) bSymbol);
+            }
+        }
+
+        Types types = Types.getInstance(new CompilerContext());
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java]: Changed

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+ */
+public class SemTypeTest {
+
+    @DataProvider(name = "filePathProvider")
+    public Object[] filePathProvider() {
+        File dir = resolvePath("test-src").toFile();
+        List<String> files = new ArrayList<>();
+        for (File file : Objects.requireNonNull(dir.listFiles())) {
+            String fileName = file.getName();
+            if (fileName.contains(FAILING_FILE) || fileName.contains(DISABLED_FILE)) {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java]: Fixed

File: tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java (hunk 3 above)
Comment (KavinduZoysa) [reply tests/jballerina-semtype-test/src/test/java/io/ballerina/test/SemTypeTest.java:125]: If we add a validation we are unable to add a different comment. So I think when we add a test we have to add the correct comment.

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**